pkg/pillar/hypervisor/pci.go

```diff
@@ -0,0 +1,200 @@
+func (r pciResource) isMem() bool {
+	return r.flags&IORESOURCE_TYPE_BITS == IORESOURCE_MEM
+}
+
+func addNoDuplicatePCI(list []typeAndPCI, tap typeAndPCI) []typeAndPCI {
+
+	for _, t := range list {
+		if t.pciLong == tap.pciLong {
+			return list
+		}
+	}
+	return append(list, tap)
+}
+
+type typeAndPCI struct {
+	pciLong string
+	ioType  types.IoType
+}
+
+// check if the PCI device is a VGA device
+// check availability of the VGA file in the sysfs filesystem
+func (t typeAndPCI) isVGA() bool {
+	bootVgaFile := sysfsPciDevices + t.pciLong + "/boot_vga"
+	if _, err := os.Stat(bootVgaFile); err == nil {
+		return true
+	}
+	return false
+}
+
+// read vendor ID
+func (t typeAndPCI) vid() (string, error) {
+	vendorID, err := os.ReadFile(sysfsPciDevices + t.pciLong + "/vendor")
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(strings.TrimSuffix(string(vendorID), "\n")), nil
+}
+
+// isBridge checks if the given PCI device is a bridge.
+// It reads the device's class from the sysfs filesystem and returns true if the class
+// starts with "0x06", which is the PCI base class code for bridges.
+// ./class file contains 3 bytes of class code as following
+// base_class,subclass,prog-if
+// e.g. 0x06,0x04,0x00 - PCI bridge
+// PCI_BASE_CLASS_BRIDGE is defined as 0x06 in include/linux/pci_ids.h
+func (t typeAndPCI) isBridge() (bool, error) {
+	class, err := os.ReadFile(sysfsPciDevices + t.pciLong + "/class")
+
+	if err != nil {
+		logrus.Errorf("Can't read PCI device class %s: %v\n",
+			t.pciLong, err)
+		return true, err // assume it is a bridge
+	}
+
+	if strings.HasPrefix(string(class), "0x06") {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// check if the PCI device a boot_vga device
+// read the boot_vga file from the sysfs filesystem
+// and return true if the file contains "1"
+func (t typeAndPCI) isBootVGA() (bool, error) {
+	bootVGA, err := os.ReadFile(sysfsPciDevices + t.pciLong + "/boot_vga")
+
+	if err != nil {
+		logrus.Errorf("Can't read PCI device boot_vga %s: %v\n",
+			t.pciLong, err)
+		return false, err
+	}
+
+	if strings.TrimSpace(string(bootVGA)) == "1" {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// readResources reads all resources of the PCI device in a list of structs {start, end, flags}
+func (t typeAndPCI) readResources() ([]pciResource, error) {
+	var resources []pciResource
+	var resourceIndexes []int
+
+	contains := func(slice []int, item int) bool {
+		for _, s := range slice {
+			if s == item {
+				return true
+			}
+		}
+		return false
+	}
+
+	path := sysfsPciDevices + t.pciLong
+
+	// read directory for device and collect valid resource indexes
+	files, err := os.ReadDir(path)
+	if err != nil {
+		return nil, logError("Can't read PCI device directory %s: %v\n",
+			path, err)
+	}
+
+	// collect indexes of valid resources
+	for _, file := range files {
+		name := file.Name()
+		if strings.HasPrefix(name, "resource") && !strings.HasSuffix(name, "_wc") {
```

Comment:
Why are we not interested in the `_wc` files? As I understand, they are exactly the same resources but mapped with the support of write-combining. It's platform-specific stuff.

Reply:
because they seem to be not involved in overhead. My formula now counts the total size of MMIO and they map the same addresses really. Do you think that a separate page table is created for the same memory region ?

Reply:
So, you mean that you see exactly the same addresses both in the resourceN and resourceN_wc files?
If it's the case, then okay, I understand why you decided to ignore one of them =)

Reply:
yes, same addresses

Reply:
If you are sure, any write-combined resource has a corresponding mapping in a regular resource - it's okay. Otherwise, we could do some kind of "`| sort | uniq`" logic: read from both lists of files and leave only unique regions.